#161966897 Implement test cases for user authentication

rkemmy opened the pull request:
#### What does this PR do?
Add user authentication tests.

#### Description of Task to be completed?
* Create a test for user registration
* Test user can reset password
* Generate user token
* Test login endpoint 
* Test social authentication
* Configure code to pep8 standards


#### How should this be manually tested?
* `git clone -b ch-user-registration-tests-161966897 https://github.com/andela/ah-infinity-stones` and ` cd ah-infinity-stones`.
* Run `python3 -m venv virtual` to create a virtual environment and `source virtual/bin/activate` to activate it.
* `pip install -r requirements.txt` to install dependencies
* Create a test database in the postgres shell
* Set up environment variables using the sample below as a guide:
```export DATABASE_USER='<postgres username>'
export DATABASE_PASSWORD='<postgres user's password>'
export HOST='localhost'
export PORT='5432'
export DATABASE_TEST='<name_of_test_database>' 
```
* Run `python3 manage.py makemigrations` followed by `python3 manage.py migrate` to apply migrations
* Run `python3 manage.py test` in the terminal to run the tests.

#### What are the relevant pivotal tracker stories?
[#161966897](https://www.pivotaltracker.com/story/show/161966897)

[Delivers #161966897]

#### Screenshot
<img width="1247" alt="screenshot 2018-12-10 at 23 23 18" src="https://user-images.githubusercontent.com/17976289/49759487-f0ea0380-fcd2-11e8-865a-ab966272cf64.png">

codeplus254 commented on authors/apps/authentication/tests/test_registration.py:
```diff
@@ -0,0 +1,262 @@
+                         "message"], "Username is a required field")
+
+    def test_missing_email(self):
+        ''' Test that checks if the required email field exists '''
```
Please remember to change the single quotes into doubles for uniformity with your other files.

rkemmy replied:
I will change that. Thanks for pointing it out

codeplus254 commented on authors/apps/authentication/models.py:
```diff
@@ -33,21 +29,21 @@ def create_user(self, username, email, password=None):
-          raise TypeError('Superusers must have a password.')
+        Superuser powers means that this use is an admin that can do anything
+        they want.
+        """
```
Great work! Could you take me through your thought process of having a superuser 👍

rkemmy replied:
A superuser is an admin who has access to the admin panel and has the power to delete offensive articles, comments, and users who do not adhere to the terms and conditions. The code you have tagged already came with the project so you can go through it and if there is anything you don't understand then you will let me know. I only used autopep8 to change it because the code was not properly indented and my linter flagged it.

codeplus254 replied:
Oh!Great!

r-wambui commented on authors/apps/authentication/tests/test_registration.py:
```diff
@@ -0,0 +1,253 @@
+from rest_framework import status
+from rest_framework.test import APIClient
+from .test_setup import BaseSetUp
+# from ..models import User
```
@rkemmy why do we have this commented code?

r-wambui commented on authors/apps/authentication/tests/test_registration.py:
```diff
@@ -0,0 +1,253 @@
+            data=json.dumps(self.duplicate_username2),
+            format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
This should be a 409 status code which indicates conflict.

r-wambui commented on authors/apps/authentication/tests/test_registration.py:
```diff
@@ -0,0 +1,253 @@
+            data=json.dumps(self.duplicate_email2),
+            format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Check the status code, research more on 409.

r-wambui commented on authors/apps/authentication/tests/test_reset_password.py:
```diff
@@ -0,0 +1,106 @@
+
+        response = self.client.put(
+            'api/users/forgot_password',
+            data=json.dumps(self.renew_password),
```
Should we have this endpoint for renew password?

r-wambui replied:
Also nothing on this test shows that the user has logged in first before trying to change password.

rkemmy committed:
chore(authentication): Tests for user authentication

    -Test user registration
    -Test user can login
    -Test user can reset password
    -Test social authentication

    [Delivers #161966897]

ThaDeveloper commented:
Great work fixing the merge conflicts. Two things to note; authentication/serializers.py need a blank line at the end of file, if you created pip with python3 you need to use pip3 (check PR).
